xbmc/XBDateTime.cpp

```diff
@@ -98,122 +72,75 @@ CDateTimeSpan CDateTimeSpan::operator -(const CDateTimeSpan& right) const
 {
   CDateTimeSpan left(*this);
 
-  LARGE_INTEGER timeLeft;
-  left.ToLargeInt(timeLeft);
-
-  LARGE_INTEGER timeRight;
-  right.ToLargeInt(timeRight);
-
-  timeLeft.QuadPart-=timeRight.QuadPart;
-
-  left.FromLargeInt(timeLeft);
+  left.m_timeSpan -= right.m_timeSpan;
 
   return left;
 }
 
 const CDateTimeSpan& CDateTimeSpan::operator +=(const CDateTimeSpan& right)
 {
-  LARGE_INTEGER timeThis;
-  ToLargeInt(timeThis);
-
-  LARGE_INTEGER timeRight;
-  right.ToLargeInt(timeRight);
-
-  timeThis.QuadPart+=timeRight.QuadPart;
-
-  FromLargeInt(timeThis);
+  m_timeSpan += right.m_timeSpan;
 
   return *this;
 }
 
 const CDateTimeSpan& CDateTimeSpan::operator -=(const CDateTimeSpan& right)
 {
-  LARGE_INTEGER timeThis;
-  ToLargeInt(timeThis);
-
-  LARGE_INTEGER timeRight;
-  right.ToLargeInt(timeRight);
-
-  timeThis.QuadPart-=timeRight.QuadPart;
-
-  FromLargeInt(timeThis);
+  m_timeSpan -= right.m_timeSpan;
 
   return *this;
 }
 
-void CDateTimeSpan::ToLargeInt(LARGE_INTEGER& time) const
-{
-  time.u.HighPart = m_timeSpan.highDateTime;
-  time.u.LowPart = m_timeSpan.lowDateTime;
-}
-
-void CDateTimeSpan::FromLargeInt(const LARGE_INTEGER& time)
-{
-  m_timeSpan.highDateTime = time.u.HighPart;
-  m_timeSpan.lowDateTime = time.u.LowPart;
-}
-
 void CDateTimeSpan::SetDateTimeSpan(int day, int hour, int minute, int second)
 {
-  LARGE_INTEGER time;
-  ToLargeInt(time);
-
-  time.QuadPart= static_cast<long long>(day) *SECONDS_PER_DAY*SECONDS_TO_FILETIME;
-  time.QuadPart+= static_cast<long long>(hour) *SECONDS_PER_HOUR*SECONDS_TO_FILETIME;
-  time.QuadPart+= static_cast<long long>(minute) *SECONDS_PER_MINUTE*SECONDS_TO_FILETIME;
-  time.QuadPart+= static_cast<long long>(second) *SECONDS_TO_FILETIME;
-
-  FromLargeInt(time);
+  m_timeSpan = KODI::TIME::floor<std::chrono::seconds>(KODI::TIME::days(day)) +
+               KODI::TIME::floor<std::chrono::seconds>(std::chrono::hours(hour)) +
+               KODI::TIME::floor<std::chrono::seconds>(std::chrono::minutes(minute)) +
+               KODI::TIME::floor<std::chrono::seconds>(std::chrono::seconds(second));
 }
 
 void CDateTimeSpan::SetFromTimeString(const std::string& time) // hh:mm
 {
   if (time.size() >= 5 && time[2] == ':')
   {
-    int hour    = atoi(time.substr(0, 2).c_str());
-    int minutes = atoi(time.substr(3, 2).c_str());
-    SetDateTimeSpan(0,hour,minutes,0);
+    int hour = std::stoi(time.substr(0, 2));
+    int minutes = std::stoi(time.substr(3, 2));
```

Comment:
`stoi `can raise exceptions when the value is malformed or out of range,
`std::from_chars` can be a solution

Reply:
Thanks for the spot on! CDateTime setters have bool result as a way to signal about failures but CDateTimeSpan does not have and I will fix that first. Then it becomes irrelevant whether to use try-catch blocks or stol because the preferred failure mode is to rrturn false and handle it at the higher logic.